core/src/main/java/net/opentsdb/query/processor/groupby/GroupByNumericArrayIterator.java

```diff
@@ -77,10 +78,78 @@
 
   protected static final int NUM_THREADS = 8;
 
+  protected static final int MAX_TS_PER_JOB = 10_000;
+
   protected static ExecutorService executorService;
 
+  protected static Thread[] threads;
+
+  public static abstract class GroupByJob<Combiner> implements Runnable {
+    private int totalTsCount;
+    private final List<TimeSeries> tsList;
+    private final int startIndex;
+    private final int endIndex;
+    private final Combiner combiner;
+    private final CountDownLatch doneSignal;
+    private StatsCollector statsCollector;
+    private final long s = System.nanoTime();
+
+    public GroupByJob(
+        int totalTsCount,
+        List<TimeSeries> tsList,
+        int startIndex,
+        int endIndex,
+        Combiner combiner,
+        CountDownLatch doneSignal,
+        StatsCollector statsCollector) {
+      this.totalTsCount = totalTsCount;
+      this.tsList = tsList;
+      this.startIndex = startIndex;
+      this.endIndex = endIndex;
+      this.combiner = combiner;
+      this.doneSignal = doneSignal;
+      this.statsCollector = statsCollector;
+      statsCollector.setGauge("groupby.timeseries.count", tsList.size());
+    }
+
+    @Override
+    public void run() {
+      statsCollector.addTime("groupby.queue.wait.time", System.nanoTime() - s, ChronoUnit.NANOS);
+      for (int i = startIndex; i < endIndex; i++) {
+        doRun(tsList.get(i), combiner);
+      }
+      doneSignal.countDown();
+    }
+
+    public abstract void doRun(TimeSeries timeSeries, Combiner combiner);
+  }
+
+  protected static BigSmallLinkedBlockingQueue<GroupByJob> blockingQueue =
```

Comment:
And actually could you move this as a member of the `GroupByFactory` please? I hate statics. And we can add a thread there to periodically report the size of the queue.